00xx-account-deletion.md

```diff
@@ -0,0 +1,576 @@
+Merkle tree leaves is free or not to compute the new frontier pointer.
+
+For example if we have data `[A, B, x, E]` in the tree, the frontier would be at
+index `3` with a free list `[2]`. Removing `E` should result in a frontier at
+index `1` and an empty free list. The max-heap data structure handles that
+scenario very gracefully in worst-case O(n log n), where $n$ is the number of
+elements in the free list.  A stack structure could incur an *O(n²)*
+worst-case cost. The min-heap would not be adequate.
+
+
+**The main benefit** of having this ordered data structure is that location
+allocation does not depend on when a location has been freed anymore. Thus, when
+syncing ledgers, the free list can be recomputed while retrieving the ledger by
+adding any empty location within the fill frontier to the currently built free list heap.
+
+There is a single caveat: the comparison function for locations needs to be
+*total* for this choice to work. The implementations of `Location` in the
+codebase all have this property.
+
+#### On-disk ledger
+
+The
+[`Database`](https://github.com/MinaProtocol/mina/blob/4495af5caea5e1bb2f98f92592c065f93a586ade/src/lib/merkle_ledger/database.ml)
+module implements an on-disk ledger that backs up the in-memory data structure.
+
+Deletion support relies on the same techniques as in-memory ledgers, that is, a
+representation of a ledger as Merkle tree *and* an additional free list.
+
+The main culprit is the representation of the free list on-disk. Here we will simply store
+the data on the KV database as follows:
+- the free list is identified by the key `free_list` (like
+  [`last_account_location`](https://github.com/MinaProtocol/mina/blob/4495af5caea5e1bb2f98f92592c065f93a586ade/src/lib/merkle_ledger/database.ml#L235));
+- store the serialized encoding of the heap in the database.
+
+**Space requirements**. We argue that this is okay to have this simple scheme in
+terms of space use.  Indeed, in the worst case today, the ledger is full and we
+do need to store it in the database.  The requirement of storing the free list
+alongside the ledger have the same worst-case scenario. When deleting an account
+one is trading the storage of an account for the storage of a location, which is
+usually smaller (a single integer instead of a complex data structure).
+Therefore storing free locations does not add more storage requirements than
+today.
+
+The serialization of and array-based heap (ike `bheap`) of integers should be rather efficient.
+
+**Allocation cost**. Allocating locations from the free list now means
+1. read `free_list`
+2. deserialize the heap value;
+3. get the needed free locations;
+4. write the accounts to the free locations;
+5. reserialize the new value of the heap.
+
+**Simple improvement on batch allocations**. Furthermore, we will implement a
+decoding function where one can specify how many locations we ideally want to
```

Comment:
Presumably this isn't useful in practise: usually, you'll be grabbing the free list so that you can update it to consume a location, so you'll need to reserialise the free list once the head elements (or tail elements as currently written here) have been removed.

It's also useful to mention performance: is there any risk that a large number of free'd accounts creates a free list that slows down accesses, because inherently we're writing a larger and larger list into the database as it grows?

Reply:
We do use batch allocations in the code. To be sure, I don't know whether the code is executed in practice.

Reply:
We do, but the point is that those batch allocations immediately result in a write of the new data. That is to say: deserialising just the head(s) is usually not what you want, given that you also want the tail so that you can reserialise it.

Reply:
I tried to explain why it might be useful. We do have to serialize some data back to be written (the tail). For batch allocations, I think it's useful to be able to get `n` locations at once to avoid `n-1` read/deserialize/serialize/write` sequences.

Reply:
I agree, but at a certain point it's simpler to deserialise the whole list, take as many locations as are needed, push any new locations onto the list, and then reserialise the final list.

Reply:
I agree too. That brings the following question: Is there a scenario where we would in one call allocate mutliple and free multiple locations at the same time?